File: csharpcode/src/CSharpCode/DataStructures/LinkedList.cs
namespace CSharpCode.DataStructures
{
    public class Node<T>
    {
        public T Value { get; set; }
        public Node<T> Next { get; set; }
        public Node(T value)
        {
            Value = value;
        }

        public override string ToString()
        {
            return new
            {
                Value,
                Next
            }.ToString();
        }
    }

    // usar equals em comparações, evitar ==
    public class LinkedList<T> : ILinkedList<T>
    {
        public Node<T> Head;
        public int Length { private set; get; }
        public LinkedList()
        {
            Length = 0;
        }

        public Node<T> Add(T value)
        {
            if (Head == null)
            {
                Head = new Node<T>(value);
                Length++;
                return Head;
            }

            var node = Head;
            while (node.Next != null)
            {
                node = node.Next;
            }
            node.Next = new Node<T>(value);
            Length++;
            return node.Next;
        }

        public void Insert(T value, int index)
        {
            if (index >= 0 && index <= Length)
            {
                var node = new Node<T>(value);
                if (index == 0)
                {
                    var current = Head;
                    node.Next = current;
                    Head = node;
                }
                else
                {
                    var previous = GetElementAt(index - 1);
                    var current = previous.Next;
                    node.Next = current;
                    previous.Next = node;
                }
                Length++;
            }
        }

        public void RemoveAt(int index)
        {
            if (index >= 0 && index < Length)
            {
                var currentNode = Head;
                if (index == 0)
                {
                    Head = currentNode.Next;
                }
                else
                {
                    Node<T> previous = null;
                    for (var i = 0; i < index; i++)
                    {
                        previous = currentNode;
                        currentNode = currentNode.Next;
                    }
                    previous.Next = currentNode.Next;
                }
                Length--;
            }
        }

        public void Remove(T value)
        {
            var index = IndexOf(value);
            RemoveAt(index);
        }

        public int IndexOf(T value)
        {
            var current = Head;
            for (var i = 0; i < Length && current != null; i++)
            {
                if (current.Value.Equals(value))
                {
                    return i;
                }
                current = current.Next;
            }
            return -1;
        }

        public Node<T> GetElementAt(int index)
        {
            if (index >= 0 && index <= Length)
            {
                Node<T> node = Head;
                for (var i = 0; i < index && node != null; i++)
                {
                    node = node.Next;
                }
                return node;
            }
            return null;
        }

        public Node<T>[] Iterable()
        {
            Node<T>[] array = new Node<T>[Length];
            if (Length > 0)
            {
                var node = Head;
                for (int i = 0; i < Length; i++)
                {
                    array[i] = node;
                    node = node.Next;
                }
            }
            return array;
        }

        public string PrintList()
        {
            return Head?.ToString();
        }
    }
}

File: dartcode/DataStructures/LinkedList.dart
class Node<T> {
  T value;
  Node<T> next;

  Node(T value) {
    this.value = value;
  }

  Map<String, dynamic> toJson() => _nodeToJson(this);

  Map<String, dynamic> _nodeToJson(Node instance) => <String, dynamic>{
        'value': instance.value,
        'next': instance.next?.toJson(),
      };
}

class LinkedList<T> {
  Node<T> head;
  int length = 0;

  Node<T> add(T value) {
    if (this.head == null) {
      this.head = new Node(value);
      this.length++;
      return this.head;
    }

    var node = this.head;
    while (node.next != null) {
      node = node.next;
    }
    node.next = new Node(value);
    this.length++;
    return node.next;
  }

  void insert(T value, int index) {
    if (index >= 0 && index <= this.length) {
      var node = new Node<T>(value);
      if (index == 0) {
        var current = this.head;
        node.next = current;
        this.head = node;
      } else {
        var previous = this.getElementAt(index - 1);
        var current = previous.next;
        node.next = current;
        previous.next = node;
      }
      this.length++;
    }
  }

  void removeAt(int index) {
    if (index >= 0 && index < this.length) {
      var currentNode = this.head;
      if (index == 0) {
        this.head = currentNode.next;
      } else {
        Node previous;
        for (var i = 0; i < index; i++) {
          previous = currentNode;
          currentNode = currentNode.next;
        }
        previous.next = currentNode.next;
      }
      this.length--;
    }
  }

  void remove(T value) {
    var index = this.indexOf(value);
    this.removeAt(index);
  }

  int indexOf(T value) {
    var current = this.head;
    for (var i = 0; i < this.length && current != null; i++) {
      if (current.value == value) {
        return i;
      }
      current = current.next;
    }
    return -1;
  }

  Node<T> getElementAt(int index) {
    if (index >= 0 && index <= this.length) {
      Node<T> node = this.head;
      for (var i = 0; i < index && node != null; i++) {
        node = node.next;
      }
      return node;
    }
    return null;
  }

  int getLength() {
    return this.length;
  }

  void printList() {
    print(this.head?.toJson());
  }
}

void main(List<String> args) {
  LinkedList<int> list = new LinkedList();
  list.add(9);
  list.add(4);
  list.add(7);
  list.add(3);
  list.add(29);
  list.add(14);
  list.printList();
  print(list.getLength());
  list.removeAt(2);
  list.printList();
  print(list.getLength());
  list.removeAt(4);
  list.printList();
  print(list.getLength());
  list.removeAt(-1);
  list.printList();
  print(list.getLength());
  list.removeAt(0);
  list.printList();
  print(list.getLength());
  print("get element by index 0: ");
  print(list.getElementAt(0).toJson());
  print("get element by index 2: ");
  print(list.getElementAt(2).toJson());
  print("insert element at index 1");
  list.insert(77, 1);
  list.printList();
  list.add(1);
  list.printList();
  print("get index by value: 77");
  print(list.indexOf(77));
  print("get index by value: 17");
  print(list.indexOf(17));
  print("remove element 29");
  list.remove(29);
  list.printList();
  print("remove element 9");
  list.remove(9);
  list.printList();
}
